packages/evm/contracts/Yaho.sol

```diff
@@ -1,76 +1,206 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 pragma solidity ^0.8.17;
 
+import { IYaho } from "./interfaces/IYaho.sol";
 import { IMessageRelay } from "./interfaces/IMessageRelay.sol";
-import { IMessageDispatcher, Message } from "./interfaces/IMessageDispatcher.sol";
+import { Message } from "./interfaces/IMessageDispatcher.sol";
 import { MessageHashCalculator } from "./utils/MessageHashCalculator.sol";
+import { MessageIdCalculator } from "./utils/MessageIdCalculator.sol";
 
-contract Yaho is IMessageDispatcher, MessageHashCalculator {
-    mapping(uint256 => bytes32) public hashes;
-    uint256 private count;
+contract Yaho is IYaho, MessageHashCalculator, MessageIdCalculator {
+    bytes32 public constant MESSAGE_BHR = keccak256("MESSAGE_BHR");
+    bytes32 public constant MESSAGE_MPI = keccak256("MESSAGE_MPI");
 
-    error NoMessagesGiven(address emitter);
-    error NoMessageIdsGiven(address emitter);
-    error NoAdaptersGiven(address emitter);
-    error UnequalArrayLengths(address emitter);
+    address public immutable headerReporter;
+
+    mapping(bytes32 => bytes32) public hashes;
+
+    constructor(address headerReporter_) {
+        headerReporter = headerReporter_;
+    }
+
+    /// @dev Dispatches a message using the EIP-5164 standard, putting their into storage and emitting their contents as an event.
+    /// @param toChainId The destination chain id.
+    /// @param to The target contract.
+    /// @param data The message data.
+    /// @return messageId A message ID corresponding to the dispatched message.
+    function dispatchMessage(uint256 toChainId, address to, bytes calldata data) external returns (bytes32 messageId) {
+        (messageId, ) = _dispatchMessage(toChainId, to, data);
+    }
+
+    /// @dev Dispatches a message using the EIP-5164 standard on more chains, putting their into storage and emitting their contents as an event.
+    /// @param toChainIds  The destination chain id array.
+    /// @param tos The target contracts.
+    /// @param data The message data.
+    /// @return messageIds An array of message IDs corresponding to the dispatched message.
+    function dispatchMessages(
```

Comment:
because if you dispatch the same data on multiple chains it is as if you generate multiple messages

Reply:
Right, I get that it goes out as multiple messages.

I'm mostly wondering what the use-case here is and whether it's common enough to justify a separate entrypoint? In other words, how often would one send the exact same data to many different destinations?

Reply:
For example if you wanted to propagate a block from a source chain to many destionation chains since now blocks propagation passes through `Yaho` in order to correctly calculate the messageId based on the EIP-5164

Reply:
Ah yeah, that's fair enough justification for a distinct entrypoint.